// @ngxs/schematics/src/utils/common/lib.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const packageJson = require('../../../package.json');
var LIBRARIES;
(function (LIBRARIES) {
    LIBRARIES["STORE"] = "@ngxs/store";
    LIBRARIES["LOGGER"] = "@ngxs/logger-plugin";
    LIBRARIES["DEVTOOLS"] = "@ngxs/devtools-plugin";
    LIBRARIES["SCHEMATICS"] = "@ngxs/schematics";
})(LIBRARIES = exports.LIBRARIES || (exports.LIBRARIES = {}));
var NodeDependencyType;
(function (NodeDependencyType) {
    NodeDependencyType["Default"] = "dependencies";
    NodeDependencyType["Dev"] = "devDependencies";
    NodeDependencyType["Peer"] = "peerDependencies";
    NodeDependencyType["Optional"] = "optionalDependencies";
})(NodeDependencyType = exports.NodeDependencyType || (exports.NodeDependencyType = {}));
exports.LIB_CONFIG = [
    {
        type: NodeDependencyType.Default,
        name: LIBRARIES.DEVTOOLS,
        version: '^3.3.2',
        overwrite: true
    },
    {
        type: NodeDependencyType.Default,
        name: LIBRARIES.LOGGER,
        version: '^3.3.2',
        overwrite: true
    },
    {
        type: NodeDependencyType.Default,
        name: LIBRARIES.STORE,
        version: '^3.3.2',
        overwrite: true
    },
    {
        type: NodeDependencyType.Dev,
        name: LIBRARIES.SCHEMATICS,
        version: `^${packageJson.version}`,
        overwrite: true
    }
];
