// @angular/compiler-cli/src/ngtsc/resource_loader.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/resource_loader", ["require", "exports", "tslib", "fs", "typescript"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var fs = require("fs");
    var ts = require("typescript");
    /**
     * `ResourceLoader` which delegates to a `CompilerHost` resource loading method.
     */
    var HostResourceLoader = /** @class */ (function () {
        function HostResourceLoader(resolver, loader) {
            this.resolver = resolver;
            this.loader = loader;
            this.cache = new Map();
            this.fetching = new Map();
        }
        HostResourceLoader.prototype.preload = function (file, containingFile) {
            var _this = this;
            var resolved = this.resolver(file, containingFile);
            if (resolved === null) {
                return undefined;
            }
            if (this.cache.has(resolved)) {
                return undefined;
            }
            else if (this.fetching.has(resolved)) {
                return this.fetching.get(resolved);
            }
            var result = this.loader(resolved);
            if (typeof result === 'string') {
                this.cache.set(resolved, result);
                return undefined;
            }
            else {
                var fetchCompletion = result.then(function (str) {
                    _this.fetching.delete(resolved);
                    _this.cache.set(resolved, str);
                });
                this.fetching.set(resolved, fetchCompletion);
                return fetchCompletion;
            }
        };
        HostResourceLoader.prototype.load = function (file, containingFile) {
            var resolved = this.resolver(file, containingFile);
            if (resolved === null) {
                throw new Error("HostResourceLoader: could not resolve " + file + " in context of " + containingFile + ")");
            }
            if (this.cache.has(resolved)) {
                return this.cache.get(resolved);
            }
            var result = this.loader(resolved);
            if (typeof result !== 'string') {
                throw new Error("HostResourceLoader: loader(" + resolved + ") returned a Promise");
            }
            this.cache.set(resolved, result);
            return result;
        };
        return HostResourceLoader;
    }());
    exports.HostResourceLoader = HostResourceLoader;
    /**
     * `ResourceLoader` which directly uses the filesystem to resolve resources synchronously.
     */
    var FileResourceLoader = /** @class */ (function () {
        function FileResourceLoader(host, options) {
            this.host = host;
            this.options = options;
        }
        FileResourceLoader.prototype.load = function (file, containingFile) {
            // Attempt to resolve `file` in the context of `containingFile`, while respecting the rootDirs
            // option from the tsconfig. First, normalize the file name.
            var e_1, _a;
            // Strip a leading '/' if one is present.
            if (file.startsWith('/')) {
                file = file.substr(1);
            }
            // Turn absolute paths into relative paths.
            if (!file.startsWith('.')) {
                file = "./" + file;
            }
            // TypeScript provides utilities to resolve module names, but not resource files (which aren't
            // a part of the ts.Program). However, TypeScript's module resolution can be used creatively
            // to locate where resource files should be expected to exist. Since module resolution returns
            // a list of file names that were considered, the loader can enumerate the possible locations
            // for the file by setting up a module resolution for it that will fail.
            file += '.$ngresource$';
            // clang-format off
            var failedLookup = ts.resolveModuleName(file, containingFile, this.options, this.host);
            // clang-format on
            if (failedLookup.failedLookupLocations === undefined) {
                throw new Error("Internal error: expected to find failedLookupLocations during resolution of resource '" + file + "' in context of " + containingFile);
            }
            var candidateLocations = failedLookup.failedLookupLocations
                .filter(function (candidate) { return candidate.endsWith('.$ngresource$.ts'); })
                .map(function (candidate) { return candidate.replace(/\.\$ngresource\$\.ts$/, ''); });
            try {
                for (var candidateLocations_1 = tslib_1.__values(candidateLocations), candidateLocations_1_1 = candidateLocations_1.next(); !candidateLocations_1_1.done; candidateLocations_1_1 = candidateLocations_1.next()) {
                    var candidate = candidateLocations_1_1.value;
                    if (fs.existsSync(candidate)) {
                        return fs.readFileSync(candidate, 'utf8');
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (candidateLocations_1_1 && !candidateLocations_1_1.done && (_a = candidateLocations_1.return)) _a.call(candidateLocations_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            throw new Error("Could not find resource " + file + " in context of " + containingFile);
        };
        return FileResourceLoader;
    }());
    exports.FileResourceLoader = FileResourceLoader;
});
